pkg/traefik/traefik.go

```diff
@@ -0,0 +1,310 @@
+
+	// Create the Traefik deployment using the service account
+	traefikDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      deploymentName,
+			Namespace: t.K8s.Namespace(),
+			Labels:    map[string]string{appLabel: appLabelValue},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: ptr.To[int32](replicas),
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{appLabel: appLabelValue},
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{appLabel: appLabelValue},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: serviceAccountName,
+					Containers: []v1.Container{
+						{
+							Name:  containerName,
+							Image: image,
+							Args: []string{
+								"--api.insecure=true",
+								"--providers.kubernetesIngress",
+								"--providers.kubernetesCRD",
+								fmt.Sprintf("--entrypoints.web.Address=:%d", Port),
+								fmt.Sprintf("--entrypoints.websecure.Address=:%d", PortSecure),
+							},
+							Ports: []v1.ContainerPort{
+								{ContainerPort: Port, Name: "web"},
+								{ContainerPort: PortSecure, Name: "websecure"},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	_, err = t.K8s.Clientset().AppsV1().Deployments(t.K8s.Namespace()).Create(ctx, traefikDeployment, metav1.CreateOptions{})
+	if err != nil {
+		return ErrTraefikDeploymentCreationFailed.Wrap(err)
+	}
+
+	if err := t.K8s.WaitForDeployment(ctx, deploymentName); err != nil {
+		return err
+	}
+
+	if err := t.createService(ctx); err != nil {
+		return err
+	}
+
+	if err := t.K8s.WaitForService(ctx, traefikServiceName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *Traefik) IP(ctx context.Context) (string, error) {
+	if t.K8s == nil {
+		return "", ErrTraefikClientNotInitialized
+	}
+
+	return t.K8s.GetServiceIP(ctx, traefikServiceName)
+}
+
+func (t *Traefik) URL(ctx context.Context, prefix string) (string, error) {
+	if t.endpoint == "" {
+		var err error
+		if t.endpoint, err = t.Endpoint(ctx); err != nil {
+			return "", ErrTraefikIPNotFound.Wrap(err)
+		}
+	}
+	return fmt.Sprintf("http://%s/%s", t.endpoint, prefix), nil
+}
+
+func (t *Traefik) Endpoint(ctx context.Context) (string, error) {
+	if t.K8s == nil {
+		return "", ErrTraefikClientNotInitialized
+	}
+	return t.K8s.GetServiceEndpoint(ctx, traefikServiceName)
+}
+
+func (t *Traefik) AddHost(ctx context.Context, serviceName, prefix string, portTCP int) error {
+	middlewareName, err := names.NewRandomK8("strip-" + prefix)
+	if err != nil {
+		return ErrGeneratingRandomK8sName.Wrap(err)
+	}
+
+	// middleware is required to strip the prefix from the service name
+	if err := t.createMiddleware(ctx, prefix, middlewareName); err != nil {
+		return err
+	}
+
+	return t.createIngressRoute(ctx, serviceName, prefix, middlewareName, portTCP)
+}
+
+// TODO: need to update the k8s pkg to handle service creation in more custom way
+func (t *Traefik) createService(ctx context.Context) error {
```

Comment:
If we deploy trafik as a knuu instance, this should be created automatically when ports are added.
The only thing we need to add is a way to make the Service a LoadBalancer type.